core/solver/idr.cpp

```diff
@@ -0,0 +1,287 @@
+}
+
+
+template <typename ValueType>
+std::unique_ptr<LinOp> Idr<ValueType>::conj_transpose() const
+{
+    return build()
+        .with_generated_preconditioner(share(
+            as<Transposable>(this->get_preconditioner())->conj_transpose()))
+        .with_criteria(this->stop_criterion_factory_)
+        .on(this->get_executor())
+        ->generate(share(
+            as<Transposable>(this->get_system_matrix())->conj_transpose()));
+}
+
+
+template <typename ValueType>
+template <typename SubspaceType>
+void Idr<ValueType>::iterate(const LinOp *b, LinOp *x) const
+{
+    using std::swap;
+    using Vector = matrix::Dense<SubspaceType>;
+    using NormVector = matrix::Dense<remove_complex<ValueType>>;
+
+    auto exec = this->get_executor();
+
+    auto one_op =
+        initialize<matrix::Dense<ValueType>>({one<ValueType>()}, exec);
+    auto neg_one_op =
+        initialize<matrix::Dense<ValueType>>({-one<ValueType>()}, exec);
+    auto subspace_neg_one_op = initialize<Vector>({-one<SubspaceType>()}, exec);
+
+    auto dense_b = as<Vector>(b);
+    auto dense_x = as<Vector>(x);
+
+    constexpr uint8 RelativeStoppingId{1};
+
+    const auto problem_size = system_matrix_->get_size()[0];
+    const auto nrhs = dense_b->get_size()[1];
+
+    auto residual = Vector::create_with_config_of(dense_b);
+    auto v = Vector::create_with_config_of(dense_b);
+    auto t = Vector::create_with_config_of(dense_b);
+    auto helper = Vector::create_with_config_of(dense_b);
+
+    auto m =
+        Vector::create(exec, gko::dim<2>{subspace_dim_, subspace_dim_ * nrhs});
+
+    auto g =
+        Vector::create(exec, gko::dim<2>{problem_size, subspace_dim_ * nrhs});
+    auto u =
+        Vector::create(exec, gko::dim<2>{problem_size, subspace_dim_ * nrhs});
+
+    auto f = Vector::create(exec, gko::dim<2>{subspace_dim_, nrhs});
+    auto c = Vector::create(exec, gko::dim<2>{subspace_dim_, nrhs});
+
+    auto omega = Vector::create(exec, gko::dim<2>{1, nrhs});
+    auto residual_norm = NormVector::create(exec, dim<2>{1, nrhs});
+    auto tht = Vector::create(exec, dim<2>{1, nrhs});
+    auto t_norm = NormVector::create(exec, dim<2>{1, nrhs});
+    auto alpha = Vector::create(exec, gko::dim<2>{1, nrhs});
+
+    bool one_changed{};
+    Array<stopping_status> stop_status(exec, nrhs);
+
+    // The dense matrix containing the randomly generated subspace vectors.
+    // Stored in column major order and complex conjugated. So, if the
+    // matrix containing the subspace vectors in row major order is called P,
+    // subspace_vectors actually contains P^H.
+    auto subspace_vectors =
+        Vector::create(exec, gko::dim<2>(subspace_dim_, problem_size));
+
+    // Initialization
+    // m = identity
+    exec->run(idr::make_initialize(m.get(), subspace_vectors.get(),
+                                   deterministic_, &stop_status));
+
+    // omega = 1
+    exec->run(
+        idr::make_fill_array(omega->get_values(), nrhs, one<SubspaceType>()));
```

Comment:
Can you not use the [`fill_array`](https://github.com/ginkgo-project/ginkgo/blob/e4b0b95a09b19de26a8af126acbb17344595e2bb/omp/components/fill_array.cpp#L43) ? Or rather why is this in the the `idr::` namespace ?

Reply:
fill_array is not in the public, so need to use it via `GKO_REGISTER_OPERATION`